presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -341,24 +354,36 @@ private PrestoSparkQueryExecution(
                 rddResults = doExecute(plan);
                 commit(session, transactionManager);
             }
-            catch (RuntimeException executionFailure) {
+            catch (Exception executionFailure) {
                 try {
                     rollback(session, transactionManager);
                 }
                 catch (RuntimeException rollbackFailure) {
-                    if (executionFailure != rollbackFailure) {
-                        executionFailure.addSuppressed(rollbackFailure);
-                    }
+                    log.error(rollbackFailure, "Encountered error when performing rollback");
                 }
+
                 try {
                     queryCompletedEvent(Optional.of(executionFailure));
                 }
                 catch (RuntimeException eventFailure) {
-                    if (executionFailure != eventFailure) {
-                        executionFailure.addSuppressed(eventFailure);
+                    log.error(eventFailure, "Error publishing query completed event");
+                }
+
+                if (executionFailure instanceof SparkException) {
```

Comment:
Instead of doing `instanceof` here, shouldn't we try to catch `SparkException` and `PrestoSparkExecutionException` in line 357? -- is that the more "Java" way to handle exceptions? 

In order to do that, you probably need to refactor line 358 to 370 into something like `tryRollbackAndLog`

Reply:
> Instead of doing instanceof here, shouldn't we try to catch SparkException and PrestoSparkExecutionException in line 357?

It feels like more just a matter of taste. We can try to catch these two exceptions separately, but then we need to make sure we replicate the rollback and query event logic correctly. We can refactor that logic into a separate method, as you suggested, but still I would prefer to have only a single block of exception handling, as it feels like it is easier to follow.